src/top-bar/selection/selection-provider.js

```diff
@@ -0,0 +1,135 @@
+import { PropTypes } from '@dhis2/prop-types'
+import React, { useEffect, useReducer } from 'react'
+import { useQueryParams, StringParam } from 'use-query-params'
+import { useCurrentUser } from '../../current-user/index.js'
+import { SelectionContext } from './selection-context.js'
+
+// TODO: should take a periodCode and return a period object
+const parsePeriodCode = code => ({
+    code: code,
+    displayName: code,
+})
+
+const initialWorkflowValue = (workflows, id) => {
+    if (id) {
+        /*
+         * Auto select workflow with query param id
+         * default to empty object if `find` returns undefined in case the
+         * workflow with the id from the url is not available to the user
+         */
+        return workflows.find(workflow => workflow.id === id) || {}
+    }
+    if (workflows.length === 1) {
+        // auto-select if user only has one workflow
+        return workflows[0]
+    }
+    return {}
+}
+
+const ACTIONS = {
+    SET_OPENED_SELECT: 'SET_OPENED_SELECT',
+    CLEAR_ALL: 'CLEAR_ALL',
+    SELECT_WORKFLOW: 'SELECT_WORKFLOW',
+    SELECT_PERIOD: 'SELECT_PERIOD',
+    SELECT_ORG_UNIT: 'SELECT_ORG_UNIT',
+}
+
+const reducer = (state, { type, payload }) => {
+    switch (type) {
+        case ACTIONS.SET_OPENED_SELECT:
+            return {
+                ...state,
+                openedSelect: payload,
+            }
+        case ACTIONS.CLEAR_ALL:
+            return {
+                openedSelect: '',
+                workflow: {},
+                period: {},
+                orgUnit: {},
+            }
+        case ACTIONS.SELECT_WORKFLOW:
+            return {
+                openedSelect: '',
+                workflow: payload,
+                period: {},
+                orgUnit: {},
+            }
+        case ACTIONS.SELECT_PERIOD:
+            return {
+                ...state,
+                openedSelect: '',
+                period: payload,
+                orgUnit: {},
+            }
+        case ACTIONS.SELECT_ORG_UNIT:
+            return {
+                ...state,
+                openedSelect: '',
+                orgUnit: payload,
+            }
+        default:
+            return state
+    }
+}
+
+const SelectionProvider = ({ children }) => {
+    const [query, setQuery] = useQueryParams({
+        wf: StringParam,
+        pe: StringParam,
+        ou: StringParam,
+    })
```

Comment:
Opinion: I don't mind having abbreviated query parameters (might even be a requirement?), but I think it makes more sense to hide the value behind a readable and/or meaningful constant.

```js
// somewhere outside the Provider copmonent
const WORKFLOW_PARAM = 'wf'
const PERIOD_PARAM = 'pe'
const ORG_UNIT_PARAM = 'ou'

// here
const [query, setQuery] = useQueryParams({
    [WORKFLOW_PARAM]: StringParam,
    [PERIOD_PARAM]: StringParam,
    [ORG_UNIT_PARAM]: StringParam,
})
```

Reply:
We discussed this part and didn't really agree. We're keeping things as they are for now.